python/cudf/cudf/core/frame.py

```diff
@@ -235,9 +241,67 @@ def _concat(cls, objs, axis=0, ignore_index=False, sort=False):
                     result_index_length += len(obj)
                     empty_has_index = empty_has_index or len(obj) > 0
 
-        # Get a list of the unique table column names
-        names = [name for f in objs for name in f._column_names]
-        names = OrderedDict.fromkeys(names).keys()
+        if join == "inner":
+            old_names = [name for obj in objs for name in obj._column_names]
+            setobjs = [set(obj._column_names) for obj in objs]
+            i = set(setobjs[0])
+            for x in setobjs[1:]:
+                i = i & set(x)
+            names_not_in_all = [name for name in old_names if name not in i]
+            names = [name for name in old_names if name in i]
+            names = OrderedDict.fromkeys(names).keys()
+            old_dtypes = [dt for obj in objs for dt in obj.dtypes]
```

Comment:
Why do we need to maintain old dtypes and type-cast to new dtypes?

Reply:
This and the other places I'm type-casting is to cater for the specific case where join='inner' and axis=1. In this case, in Pandas, all columns retain their datatype. Other than creating a new DataFrame from scratch and assigning its column names and its old dtypes, I'm not sure what other options there are to handling this case? 

I'll look through the axis=1 code again in case there's anything there I could use that would avoid this.

Reply:
Hi Prem! Just a note on the changes. I used the code similar to that under axis=1 in reshape.py to make sure the data in the new DataFrame is the same as the original one. When I do this the Dataframe retains the dtypes and I don't need to typecast anymore :D Tests seem to be passing on my end.